subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/CppCompiler.java

```diff
@@ -23,14 +23,19 @@
 import org.gradle.nativeplatform.toolchain.internal.CommandLineToolContext;
 import org.gradle.nativeplatform.toolchain.internal.CommandLineToolInvocationWorker;
 import org.gradle.nativeplatform.toolchain.internal.compilespec.CppCompileSpec;
+import org.gradle.nativeplatform.toolchain.internal.metadata.CompilerType;
+import org.gradle.util.VersionNumber;
 
 class CppCompiler extends GccCompatibleNativeCompiler<CppCompileSpec>  {
-
-    CppCompiler(BuildOperationExecutor buildOperationExecutor, CompilerOutputFileNamingSchemeFactory compilerOutputFileNamingSchemeFactory, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, String objectFileExtension, boolean useCommandFile, WorkerLeaseService workerLeaseService) {
-        super(buildOperationExecutor, compilerOutputFileNamingSchemeFactory, commandLineToolInvocationWorker, invocationContext, new CppCompileArgsTransformer(), Transformers.<CppCompileSpec>noOpTransformer(), objectFileExtension, useCommandFile, workerLeaseService);
+    CppCompiler(BuildOperationExecutor buildOperationExecutor, CompilerOutputFileNamingSchemeFactory compilerOutputFileNamingSchemeFactory, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, String objectFileExtension, boolean useCommandFile, WorkerLeaseService workerLeaseService, CompilerType compilerType, VersionNumber compilerVersion) {
+        super(buildOperationExecutor, compilerOutputFileNamingSchemeFactory, commandLineToolInvocationWorker, invocationContext, new CppCompileArgsTransformer(compilerType, compilerVersion), Transformers.<CppCompileSpec>noOpTransformer(), objectFileExtension, useCommandFile, workerLeaseService);
```

Comment:
❌ Instead of passing the version inside the constructor and causing lots of waterfall changes, I think we should simply set the version on the `spec` [here](https://github.com/gradle/gradle/blob/master/subprojects/platform-base/src/main/java/org/gradle/language/base/internal/compile/VersionAwareCompiler.java#L34). You would be required to add `CompilerVersion` getter/setter on `CompileSpec`. You will need to add test coverage for version aware compiler wrapped vs normal compiler. There isn't any yet, so just add a unit test `Specification` as `VersionAwareCompilerTest` and mock the wrapped compiler to assert the version is properly set on the `CompileSpec`.

The `CompilerVersion` most likely contains the String that would identify Clang vs GCC, however, I would still create typed `CppCompiler` for both tool chain simply because we want to avoid comparing strings and such.

Reply:
Should we really be adding this getter/setter on `CompileSpec`? It's pretty high-level interface, and it would require changing a number of classes unrelated to this PR (e.g., `JavadocSpec`).